### PR TITLE
SSCS-5104 Add draft endpoint to SYA controller

### PR DESCRIPTION
**Pull request by PaulVM**
WIP - This PR is for review purposes only.

### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/SSCS-5104

### Change description ###
Add draft endpoint and service

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/sscs/controller/SyaController.java [hunk 1]
```diff
@@ -44,4 +44,23 @@
...
         return status(201).build();
     }
+
+
+    @ApiOperation(value = "submitDraftAppeal",
+            notes = "Creates a draft case from the SYA details",
+            response = String.class, responseContainer = "Appeal details")
+    @ApiResponses(value = {@ApiResponse(code = 201, message = "Submitted draft appeal successfully",
+            response = String.class)})
+    @RequestMapping(value = "/draft", method = POST, consumes = APPLICATION_JSON_VALUE)
```
Comment (iguannago) [inline src/main/java/uk/gov/hmcts/reform/sscs/controller/SyaController.java]: should be plural /drafts accordingly API definition best practices

File: src/main/java/uk/gov/hmcts/reform/sscs/controller/SyaController.java [hunk 2]
```diff
@@ -44,4 +44,23 @@
...
+    public ResponseEntity<String> createDraftAppeal(@RequestBody SyaCaseWrapper syaCaseWrapper) {
+        log.info("Draft appeal with Nino - {} and benefit type {} received", syaCaseWrapper.getAppellant().getNino(),
```
Comment (iguannago) [inline src/main/java/uk/gov/hmcts/reform/sscs/controller/SyaController.java]: let's remove this log from now.

File: src/main/java/uk/gov/hmcts/reform/sscs/controller/SyaController.java [hunk 3]
```diff
@@ -44,4 +44,23 @@
...
+                syaCaseWrapper.getBenefitType().getCode());
+        Long caseId = submitAppealService.submitDraftAppeal(syaCaseWrapper);
+        log.info("Draft case {} with Nino - {} and benefit type - {} processed successfully",
```
Comment (iguannago) [inline src/main/java/uk/gov/hmcts/reform/sscs/controller/SyaController.java]: Let's rephrase this log wording to just log the draft ccdId for now.

File: src/main/java/uk/gov/hmcts/reform/sscs/service/SubmitAppealService.java [hunk 4]
```diff
@@ -67,6 +64,16 @@ public Long submitAppeal(SyaCaseWrapper appeal) {
         return (caseDetails != null) ? caseDetails.getId() : null;
     }
 
+    public Long submitDraftAppeal(SyaCaseWrapper appeal) {
+        SscsCaseData caseData = convertSyaToCcdCaseData(appeal);
+
+        IdamTokens idamTokens = idamService.getIdamTokens();
+        SscsCaseDetails caseDetails = createCaseInCcd(caseData, EventType.DRAFT, idamTokens);
```
Comment (iguannago) [inline src/main/java/uk/gov/hmcts/reform/sscs/service/SubmitAppealService.java]: let's duplicate this createCaseInCdd and rename something like createDraftInCcd so that you can have a specific implementation for the draft. the current createCaseInCcd takes care of duplication issues whereas the draft doesn't have to

File: src/test/java/uk/gov/hmcts/reform/sscs/service/SubmitAppealServiceTest.java [hunk 5]
```diff
@@ -275,11 +275,10 @@ public void testNullPostCode() {
...
+    public void testPrepareCaseForCcd() {
         SyaCaseWrapper appealData = getSyaCaseWrapper();
-        RegionalProcessingCenter rpc = getRegionalProcessingCenter();
-        SscsCaseData caseData = submitAppealService.transformAppealToCaseData(appealData, "Cardiff", rpc);
-        assertEquals("Cardiff", caseData.getRegion());
+        SscsCaseData caseData = submitAppealService.prepareCaseForCcd(appealData, "CF10");
+        assertEquals("CARDIFF", caseData.getRegion());
     }
 
     @Test
```
Comment (iguannago) [inline src/test/java/uk/gov/hmcts/reform/sscs/service/SubmitAppealServiceTest.java]: Please, add junit test to cover the submitDraft path as well.

File: src/main/java/uk/gov/hmcts/reform/sscs/controller/SyaController.java (hunk 1 above)
Comment (PaulVM) [reply src/main/java/uk/gov/hmcts/reform/sscs/controller/SyaController.java]: done

File: src/main/java/uk/gov/hmcts/reform/sscs/controller/SyaController.java (hunk 2 above)
Comment (PaulVM) [reply src/main/java/uk/gov/hmcts/reform/sscs/controller/SyaController.java]: removed.

File: src/main/java/uk/gov/hmcts/reform/sscs/controller/SyaController.java (hunk 3 above)
Comment (PaulVM) [reply src/main/java/uk/gov/hmcts/reform/sscs/controller/SyaController.java]: rephrased.

File: src/main/java/uk/gov/hmcts/reform/sscs/service/SubmitAppealService.java (hunk 4 above)
Comment (PaulVM) [reply src/main/java/uk/gov/hmcts/reform/sscs/service/SubmitAppealService.java]: Duplicated.

File: src/main/java/uk/gov/hmcts/reform/sscs/service/SubmitAppealService.java [hunk 6]
```diff
@@ -128,6 +134,25 @@ private SscsCaseDetails createCaseInCcd(SscsCaseData caseData, EventType eventTy
         }
     }
 
+    private SscsCaseDetails createDraftCaseInCcd(SscsCaseData caseData, EventType eventType, IdamTokens idamTokens) {
+        SscsCaseDetails caseDetails = null;
+        try {
```
[the remark this answers is not in the document]
Comment (PaulVM) [reply src/main/java/uk/gov/hmcts/reform/sscs/service/SubmitAppealService.java]: Error handling removed.

File: src/main/java/uk/gov/hmcts/reform/sscs/domain/wrapper/SyaCaseWrapper.java [hunk 7]
```diff
@@ -136,15 +136,20 @@ public void setSignAndSubmit(SyaSignAndSubmit signAndSubmit) {
     }
 
     public SyaContactDetails getContactDetails() {
+        if (null == appellant) {
```
Comment (iguannago) [inline src/main/java/uk/gov/hmcts/reform/sscs/domain/wrapper/SyaCaseWrapper.java]: TODO: Add junit test to cover different scenarios such as: only appellant is null, only appointee is null, both are null and both are populated

Comment (PaulVM) [reply src/main/java/uk/gov/hmcts/reform/sscs/domain/wrapper/SyaCaseWrapper.java]: Was this done?

Comment (iguannago) [reply src/main/java/uk/gov/hmcts/reform/sscs/domain/wrapper/SyaCaseWrapper.java]: it's not, but it's not a blocker for now and I will address as part of the next ticket if that's ok.

File: src/test/java/uk/gov/hmcts/reform/sscs/util/Helper.java [hunk 8]
```diff
@@ -0,0 +1,18 @@
+package uk.gov.hmcts.reform.sscs.util;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+public final class Helper {
```
[the remark this answers is not in the document]
Comment (iguannago) [reply src/test/java/uk/gov/hmcts/reform/sscs/util/Helper.java]: Yes, agree.  I will change it now...

File: src/main/java/uk/gov/hmcts/reform/sscs/domain/wrapper/SyaCaseWrapper.java (hunk 7 above)
Comment (PaulVM) [reply src/main/java/uk/gov/hmcts/reform/sscs/domain/wrapper/SyaCaseWrapper.java]: ok

File: src/e2e/java/uk/gov/hmcts/reform/sscs/functional/sya/SubmitDraftTest.java [hunk 9]
```diff
@@ -25,6 +27,9 @@
     @Value("${test-url}")
     private String testUrl;
 
+    @Value("${test-oauth2Token}")
+    String oauth2Token;
```
Comment (iguannago) [inline src/e2e/java/uk/gov/hmcts/reform/sscs/functional/sya/SubmitDraftTest.java]: we cannot inject a hardcoded token. This will fail in the pipeline since that user does not exist in AAT. You have to use a user that exists in AAT. We have created this one: sscs-citizen@hmcts.net. 
Check with me for the password. This user needs to be configured in the properties files.

File: src/main/java/uk/gov/hmcts/reform/sscs/config/CitizenCcdService.java [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    private SscsCcdConvertService sscsCcdConvertService;
+
+    @Autowired
+    public CitizenCcdService(CitizenCcdClient citizenCcdClient, SscsCcdConvertService sscsCcdConvertService) {
+        this.citizenCcdClient = citizenCcdClient;
+        this.sscsCcdConvertService = sscsCcdConvertService;
+    }
+
+    public SscsCaseDetails createCase(SscsCaseData caseData, String eventType, String summary, String description, IdamTokens idamTokens) {
+        log.info("Starting create case process with SC number {} and ccdID {} and eventType {} ...",
```
Comment (iguannago) [inline src/main/java/uk/gov/hmcts/reform/sscs/config/CitizenCcdService.java]: Remove logs

File: src/main/java/uk/gov/hmcts/reform/sscs/config/CitizenCcdService.java [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+                caseData.getCaseReference(), caseData.getCcdCaseId(), eventType);
+        try {
```
Comment (iguannago) [inline src/main/java/uk/gov/hmcts/reform/sscs/config/CitizenCcdService.java]: NO error handling at this stage, please.

File: src/main/java/uk/gov/hmcts/reform/sscs/config/CitizenCcdClient.java [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+import uk.gov.hmcts.reform.ccd.client.CoreCaseDataApi;
+import uk.gov.hmcts.reform.ccd.client.model.CaseDataContent;
+import uk.gov.hmcts.reform.ccd.client.model.CaseDetails;
+import uk.gov.hmcts.reform.ccd.client.model.StartEventResponse;
+import uk.gov.hmcts.reform.sscs.ccd.config.CcdRequestDetails;
+import uk.gov.hmcts.reform.sscs.idam.IdamTokens;
+
+@Service
+@Slf4j
+public class CitizenCcdClient {
```
Comment (iguannago) [inline src/main/java/uk/gov/hmcts/reform/sscs/config/CitizenCcdClient.java]: YOu need to add junit coverage for this

File: src/main/java/uk/gov/hmcts/reform/sscs/config/CitizenCcdService.java [hunk 13]
```diff
@@ -0,0 +1,54 @@
+package uk.gov.hmcts.reform.sscs.config;
```
Comment (iguannago) [inline src/main/java/uk/gov/hmcts/reform/sscs/config/CitizenCcdService.java]: Junit coverage for this.

File: src/e2e/resources/config/application_e2e.properties [hunk 14]
```diff
@@ -1,2 +1,3 @@
 #postcode
-test-url: ${TEST_URL:http://localhost:8080}
\ No newline at end of file
+test-url: ${TEST_URL:http://localhost:8080}
```
Comment (iguannago) [inline src/e2e/resources/config/application_e2e.properties]: I can configure this for you @arcotc .

File: src/main/java/uk/gov/hmcts/reform/logging/exception/AbstractLoggingException.java [hunk 15]
```diff
@@ -0,0 +1,83 @@
+package uk.gov.hmcts.reform.logging.exception;
+
+import ch.qos.logback.classic.Level;
+import ch.qos.logback.classic.spi.ThrowableProxy;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class AbstractLoggingException extends RuntimeException {
```
Comment (iguannago) [inline src/main/java/uk/gov/hmcts/reform/logging/exception/AbstractLoggingException.java]: what is this class about?

File: src/e2e/java/uk/gov/hmcts/reform/sscs/functional/sya/SubmitDraftTest.java [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+            "sscs",
+            idamOauth2RedirectUrl,
+            " "
+        );
+
+        Authorize authorizeToken = idamApiClient.authorizeToken(
+            authorize.getCode(),
+            "authorization_code",
+            idamOauth2RedirectUrl,
+            "sscs",
```
Comment (arcotc) [inline src/e2e/java/uk/gov/hmcts/reform/sscs/functional/sya/SubmitDraftTest.java]: Make this a constant so that it is obvious what it is and not just a `magic string`.

File: src/e2e/java/uk/gov/hmcts/reform/sscs/functional/sya/SubmitDraftTest.java [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    public String getIdamOauth2Token(String username, String password) {
+        String authorisation = username + ":" + password;
+        String base64Authorisation = Base64.getEncoder().encodeToString(authorisation.getBytes());
+
+        Authorize authorize = idamApiClient.authorizeCodeType(
+            "Basic " + base64Authorisation,
+            "code",
+            "sscs",
```
Comment (arcotc) [inline src/e2e/java/uk/gov/hmcts/reform/sscs/functional/sya/SubmitDraftTest.java]: Make this a constant so that it is obvious what it is and not just a `magic string`.

File: src/e2e/java/uk/gov/hmcts/reform/sscs/functional/sya/SubmitDraftTest.java [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+            .assertThat().body("id", not(isEmptyOrNullString())).log().all(true);
+    }
+
+    public String getIdamOauth2Token(String username, String password) {
+        String authorisation = username + ":" + password;
+        String base64Authorisation = Base64.getEncoder().encodeToString(authorisation.getBytes());
+
+        Authorize authorize = idamApiClient.authorizeCodeType(
+            "Basic " + base64Authorisation,
+            "code",
```
Comment (arcotc) [inline src/e2e/java/uk/gov/hmcts/reform/sscs/functional/sya/SubmitDraftTest.java]: Make this a constant so that it is obvious what it is and not just a `magic string`.

File: src/e2e/java/uk/gov/hmcts/reform/sscs/functional/sya/SubmitDraftTest.java [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+            "sscs",
+            idamOauth2RedirectUrl,
+            " "
+        );
+
+        Authorize authorizeToken = idamApiClient.authorizeToken(
+            authorize.getCode(),
+            "authorization_code",
```
Comment (arcotc) [inline src/e2e/java/uk/gov/hmcts/reform/sscs/functional/sya/SubmitDraftTest.java]: Make this a constant so that it is obvious what it is and not just a `magic string`.

File: src/main/java/uk/gov/hmcts/reform/sscs/controller/SyaController.java [hunk 20]
```diff
@@ -43,4 +47,20 @@
 
         return status(201).build();
```
Comment (arcotc) [inline src/main/java/uk/gov/hmcts/reform/sscs/controller/SyaController.java]: Should this not also return the http header `location`?  Use `HttpStatus.CREATED` instead of `201`.

File: src/main/java/uk/gov/hmcts/reform/sscs/controller/SyaController.java [hunk 21]
```diff
@@ -43,4 +47,20 @@
...
+    @PostMapping(value = "/drafts", consumes = APPLICATION_JSON_VALUE)
+    public ResponseEntity<Draft> createDraftAppeal(
+            @RequestHeader(AUTHORIZATION) String authorisation,
+            @RequestBody SyaCaseWrapper syaCaseWrapper) {
+        Preconditions.checkNotNull(syaCaseWrapper);
+        Draft draft = Draft.builder()
+                .id(submitAppealService.submitDraftAppeal(authorisation, syaCaseWrapper))
+                .build();
+        log.info("{} processed successfully", draft);
+        return status(HttpStatus.CREATED).body(draft);
```
Comment (arcotc) [inline src/main/java/uk/gov/hmcts/reform/sscs/controller/SyaController.java]: Should this not also return the http header `location`?

File: src/main/java/uk/gov/hmcts/reform/logging/exception/AbstractLoggingException.java (hunk 15 above)
Comment (iguannago) [reply src/main/java/uk/gov/hmcts/reform/logging/exception/AbstractLoggingException.java]: remove class, no needed.

File: src/e2e/java/uk/gov/hmcts/reform/sscs/functional/sya/SubmitDraftTest.java (hunk 9 above)
Comment (iguannago) [reply src/e2e/java/uk/gov/hmcts/reform/sscs/functional/sya/SubmitDraftTest.java]: addressed.

File: src/e2e/java/uk/gov/hmcts/reform/sscs/functional/sya/SubmitDraftTest.java (hunk 16 above)
Comment (iguannago) [reply src/e2e/java/uk/gov/hmcts/reform/sscs/functional/sya/SubmitDraftTest.java]: this code needs to be refactored later on, but yes, I will do it now, thanks.

File: src/e2e/java/uk/gov/hmcts/reform/sscs/functional/sya/SubmitDraftTest.java (hunk 17 above)
Comment (iguannago) [reply src/e2e/java/uk/gov/hmcts/reform/sscs/functional/sya/SubmitDraftTest.java]: done

File: src/e2e/java/uk/gov/hmcts/reform/sscs/functional/sya/SubmitDraftTest.java (hunk 18 above)
Comment (iguannago) [reply src/e2e/java/uk/gov/hmcts/reform/sscs/functional/sya/SubmitDraftTest.java]: done

File: src/e2e/java/uk/gov/hmcts/reform/sscs/functional/sya/SubmitDraftTest.java (hunk 19 above)
Comment (iguannago) [reply src/e2e/java/uk/gov/hmcts/reform/sscs/functional/sya/SubmitDraftTest.java]: done

File: src/main/java/uk/gov/hmcts/reform/sscs/controller/SyaController.java (hunk 21 above)
Comment (iguannago) [reply src/main/java/uk/gov/hmcts/reform/sscs/controller/SyaController.java]: done
